### PR TITLE
[api] support actions from ledger with eip-2718

### LINKED ISSUE #2916: [api] add web3 ledger support on native babel
**What would you like to be added**:
https://github.com/iotexproject/iotex-antenna/commit/8a44c1ba587499ba684ac4817ce49f8a2a282868
**Why is this needed**:

### PR DESCRIPTION
**Pull request by Liuhaai**
fix #2916

### CODE REVIEW

**Review by ququzone — APPROVED** [review]

File: api/web3server.go [hunk 1]
```diff
@@ -436,7 +436,13 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
 		return nil, err
 	}
-	tx, sig, pubkey, err := action.DecodeRawTx(dataStr, svr.coreService.EVMNetworkID())
+
+	chainiD := svr.coreService.EVMNetworkID()
+	// tx from ledger is signed with chainID 999999
+	if isSpecialTx(dataStr, chainiD) {
+		chainiD = 999999
+	}
+	tx, sig, pubkey, err := action.DecodeRawTx(dataStr, chainiD)
```
Comment (dustinxie) [inline api/web3server.go:445]:
```
action.DecodeRawTx(dataStr, svr.coreService.EVMNetworkID())
if err != nil {
   // ledger sign with 999999
   action.DecodeRawTx(dataStr, 999999)
}
```
chainId = ExtractChainID(dataStr, svr.coreService.EVMNetworkID())

File: api/web3server_utils.go [hunk 2]
```diff
@@ -352,6 +355,22 @@ func (svr *Web3Server) isContractAddr(addr string) (bool, error) {
 	return accountMeta.IsContract, nil
 }
 
+func isSpecialTx(rawData string, chainID uint32) bool {
+	dataInString, err := hex.DecodeString(util.Remove0xPrefix(rawData))
```
Comment (dustinxie) [inline api/web3server_utils.go:359]: Remove0xPrefix --> action/ or pkg/

File: action/rlp_tx.go [hunk 3]
```diff
@@ -84,33 +85,70 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
 // DecodeRawTx decodes raw data string into eth tx
 func DecodeRawTx(rawData string, chainID uint32) (tx *types.Transaction, sig []byte, pubkey crypto.PublicKey, err error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:86]: do not use named return

File: action/rlp_tx.go [hunk 4]
```diff
@@ -84,33 +85,70 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
+	}
+
+	// recover public key
+	rawHash := types.NewEIP155Signer(big.NewInt(int64(chainID))).Hash(tx)
+	pubkey, err = crypto.RecoverPubkey(rawHash[:], sig)
+	return
+}
+
+// EncodeRawTx encodes action into the data string of eth tx
+func EncodeRawTx(act Action, pvk crypto.PrivateKey, chainID uint32) (rawData string, err error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:112]: same

File: action/rlp_tx.go [hunk 5]
```diff
@@ -82,35 +83,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
+	tx := &types.Transaction{}
+	if err = rlp.DecodeBytes(dataInString, tx); err != nil {
+		return 0, err
+	}
+
+	v, _, _ := tx.RawSignatureValues()
+	recID := uint32(v.Int64()) - 2*chainID - 8
+
+	// tx from ledger is signed with chainID 999999
+	if recID != 27 && recID != 28 {
```
Comment (dustinxie) [inline action/rlp_tx.go:157]:
```
if recID >= 27 && recID <= 30 {
    return chainID, nil
}
// tx from ledger is signed with chainID 999999
recID := uint32(v.Int64()) - 2*999999 - 8
if recID >= 27 && recID <= 30 {
    return chainID, nil
}
return 0, err
```

File: action/rlp_tx.go [hunk 6]
```diff
@@ -82,35 +83,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
+	}
+	encodedTx, err := rlp.EncodeToBytes(signedTx)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(encodedTx[:]), nil
+}
+
+// HandleChainID handles special tx from web3 which signed by ledger with chainID 999999
+func HandleChainID(rawData string, chainID uint32) (uint32, error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:142]: DecodeChainID

File: api/web3server_utils.go [hunk 7]
```diff
@@ -24,7 +24,7 @@ import (
 	"google.golang.org/grpc/status"
 
 	logfilter "github.com/iotexproject/iotex-core/api/logfilter"
-	"github.com/iotexproject/iotex-core/ioctl/util"
+	ioctlUtil "github.com/iotexproject/iotex-core/ioctl/util"
```
Comment (dustinxie) [inline api/web3server_utils.go:27]: revert this file change

File: api/web3server.go [hunk 8]
```diff
@@ -436,7 +436,12 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	tx, sig, pubkey, err := action.DecodeRawTx(dataStr, svr.coreService.EVMNetworkID())
+
+	chainiD, err := action.HandleChainID(dataStr, svr.coreService.EVMNetworkID())
+	if err != nil {
+		return nil, err
+	}
+	tx, sig, pubkey, err := action.DecodeRawTx(dataStr, chainiD)
```
Comment (dustinxie) [inline api/web3server.go:444]: for chainID = 999999, the internal payload was actually native protobuf, so need to recover pubkey that way
see: https://github.com/iotexproject/iotex-antenna/blob/ledger-rlp/src/iotx.ts#L161-L178

File: action/rlp_tx.go [hunk 9]
```diff
@@ -82,35 +87,150 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
+	}
+	encodedTx, err := rlp.EncodeToBytes(signedTx)
+	if err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(encodedTx[:]), nil
+}
+
+// ExtractSignatureAndPubkey calculates signature and pubkey from tx
+func ExtractSignatureAndPubkey(tx *types.Transaction, pbAct *iotextypes.ActionCore, chainID uint32, isEthEncoding bool) ([]byte, crypto.PublicKey, error) {
```
Comment (CoderZhi) [inline action/rlp_tx.go:133]: wrong place to handle the logic. 
1. This is supposed to be at the api level
2. signature has nothing to do with pbAct

File: api/web3server.go [hunk 10]
```diff
@@ -458,15 +459,19 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
 			GasPrice: gasPrice,
-			ChainID:  svr.coreService.ChainID(),
+			// TODO: change the chainid to svr.coreService.ChainID()
+			ChainID: 0,
 		},
-		SenderPubKey: pubkey.Bytes(),
-		Signature:    sig,
-		Encoding:     iotextypes.Encoding_ETHEREUM_RLP,
+	}
+
```
Comment (CoderZhi) [inline api/web3server.go:470]:
```
    req.Signature, err := getSignatureFromRLPTX(tx, chainID)
    if err != nil {
        return nil, err
    }
```

File: api/web3server.go [hunk 11]
```diff
@@ -458,15 +459,19 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
+	if isEthEncoding {
+		req.Encoding = iotextypes.Encoding_ETHEREUM_RLP
```
Comment (CoderZhi) [inline api/web3server.go:472]:
```
h := types.NewEIP155Signer(big.NewInt(int64(chainID))).Hash(tx)
rawHash = h[:]
```

File: api/web3server.go [hunk 12]
```diff
@@ -458,15 +459,19 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
+	} else {
+		req.Encoding = iotextypes.Encoding_IOTEX_PROTOBUF
```
Comment (CoderZhi) [inline api/web3server.go:474]: is the encoding related to protobuf?

File: api/web3server.go [hunk 13]
```diff
@@ -458,15 +459,19 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
 	}
```
Comment (CoderZhi) [inline api/web3server.go:475]: move to line 498

File: action/rlp_tx.go [hunk 14]
```diff
@@ -82,35 +81,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
-	return
+	legacyTxWithEncodingType struct {
+		// data fields of legacy transaction.
+		Nonce    uint64
+		GasPrice *big.Int
+		Gas      uint64
+		To       *common.Address `rlp:"nil"`
+		Value    *big.Int
+		Data     []byte
+		V, R, S  *big.Int
```
Comment (dustinxie) [inline action/rlp_tx.go:130]: can directly use `type.LegacyTx`?

File: action/rlp_tx.go [hunk 15]
```diff
@@ -82,35 +81,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
 	copy(sig[32-rSize:32], r.Bytes())
 	sSize := len(s.Bytes())
 	copy(sig[64-sSize:], s.Bytes())
 	sig = append(sig, byte(recID))
+	return sig, nil
+}
+
+type (
+	// transactionUnwrapper is a unwrapper for the Ethereum transaction.
+	transactionUnwrapper struct {
```
Comment (dustinxie) [inline action/rlp_tx.go:117]: `txUnwrapper` for simplicity?

File: action/rlp_tx.go [hunk 16]
```diff
@@ -82,35 +81,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
...
+		Value:    tx.Value,
+		Data:     tx.Data,
+		V:        tx.V,
+		R:        tx.R,
+		S:        tx.S,
+	}
+}
+
+func (tx *legacyTxWithEncodingType) isEthEncoding() bool {
+	if len(tx.EncodingType) > 0 && hex.EncodeToString(tx.EncodingType) == "01" {
```
Comment (dustinxie) [inline action/rlp_tx.go:171]: pending Leo's final implementation, i think we discussed using a 4-byte magic number

File: action/rlp_tx.go [hunk 17]
```diff
@@ -82,35 +81,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
 }
 
 // DecodeRawTx decodes raw data string into eth tx
-func DecodeRawTx(rawData string, chainID uint32) (tx *types.Transaction, sig []byte, pubkey crypto.PublicKey, err error) {
+func DecodeRawTx(rawData string, chainID uint32) (*types.Transaction, bool, error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:84]: why take out `sig, pubkey`, i think we should keep it?

File: api/web3server.go [hunk 18]
```diff
@@ -438,7 +443,9 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	tx, sig, pubkey, err := action.DecodeRawTx(dataStr, svr.coreService.EVMNetworkID())
```
Comment (dustinxie) [inline api/web3server.go:441]: we should keep it like this

File: api/web3server.go [hunk 19]
```diff
@@ -491,6 +496,29 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
+
+	var rawHash []byte
+	if isEthEncoding {
+		req.Encoding = iotextypes.Encoding_ETHEREUM_RLP
+		h := types.NewEIP155Signer(big.NewInt(int64(chainID))).Hash(tx)
+		rawHash = h[:]
+	} else {
+		req.Encoding = iotextypes.Encoding_IOTEX_PROTOBUF
+		h := hash.Hash256b(byteutil.Must(proto.Marshal(req.Core)))
+		rawHash = h[:]
```
Comment (dustinxie) [inline api/web3server.go:512]: keep the above `tx, sig, pubkey, err := action.DecodeRawTx()`
so in web3server we don't need to handle these details, they belong to action package

File: api/web3server.go [hunk 20]
```diff
@@ -458,15 +465,13 @@ func (svr *Web3Server) sendRawTransaction(in interface{}) (interface{}, error) {
...
 	req := &iotextypes.Action{
 		Core: &iotextypes.ActionCore{
-			Version:  0,
+			Version:  1,
 			Nonce:    tx.Nonce(),
 			GasLimit: tx.Gas(),
 			GasPrice: gasPrice,
-			ChainID:  svr.coreService.ChainID(),
+			// TODO: change the chainid to svr.coreService.ChainID()
+			ChainID: 0,
```
Comment (dustinxie) [inline api/web3server.go:473]: we don't need to change this? i mean, using `svr.coreService.ChainID()` works

**Comment by Liuhaai** [comment]
Due to the change of my account, the pr will be submitted in a new one.

**Pull request closed without merging**